### PR TITLE
task/34162 install devise gem and cofig

### PR DESCRIPTION
**Pull request by phuongbk**

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,5 +1,11 @@
...
-  # Prevent CSRF attacks by raising an exception.
-  # For APIs, you may want to use :null_session instead.
+  before_action :configure_permitted_parameters, if: :devise_controller?
+
   protect_from_forgery with: :exception
+
+  protected
+
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.for(:sign_up) {|u| u.permit(:name, :email, :password, :password_confirmation)}
```
Comment (vuducdat) [inline app/controllers/application_controller.rb]: do you sure this sentence <= 85 character ?

File: app/models/user.rb [hunk 2]
```diff
@@ -1,4 +1,6 @@
 class User < ActiveRecord::Base
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable
```
Comment (vuducdat) [inline app/models/user.rb]: you should enter ?

File: app/views/devise/registrations/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,38 @@
+<% provide :title, t(".title") %>
+<h1><%= t ".title" %></h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(resource, as: resource_name, url: registration_path(resource_name)) do |f| %>
```
Comment (yuyuvn) [inline app/views/devise/registrations/new.html.erb]: convention

Comment (vuducdat) [inline app/views/devise/registrations/new.html.erb]: do you sure this sentence <= 85 character ?

File: app/views/devise/registrations/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+      <div class="field">
+        <%= f.label :email %><br />
+        <%= f.email_field :email %>
+      </div>
+
+      <div class="field">
+        <%= f.label :password %>
+        <% if @minimum_password_length %>
+        <em>(<%= @minimum_password_length %> characters minimum)</em>
```
Comment (yuyuvn) [inline app/views/devise/registrations/new.html.erb]: indent

File: app/views/devise/sessions/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,22 @@
+<% provide :title, t(".title") %>
+<h1><%= t ".title" %></h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(resource, as: resource_name, url: session_path(resource_name)) do |f| %>
```
Comment (vuducdat) [inline app/views/devise/sessions/new.html.erb]: do you sure this sentence <= 85 character ?

File: app/views/devise/registrations/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        <%= f.label :password_confirmation %><br />
+        <%= f.password_field :password_confirmation, autocomplete: "off" %>
+      </div>
+
+      <div class="actions">
+        <%= f.submit t(".submit-btn"), class: "btn btn-primary" %>
+      </div>
+    <% end %>
+
+    <%= render "devise/shared/links" %>
```
Comment (yuyuvn) [inline app/views/devise/registrations/new.html.erb]: remove

File: app/views/devise/sessions/new.html.erb (hunk 5 above)
Comment (yuyuvn) [inline app/views/devise/sessions/new.html.erb]: convention

File: app/views/devise/sessions/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+      <div class="field">
+        <%= f.label :password %><br />
+        <%= f.password_field :password, autocomplete: "off" %>
+      </div>
+
+      <div class="actions">
+        <%= f.submit t(".submit-btn"), class: "btn btn-primary" %>
+      </div>
+    <% end %>
+    <%= render "devise/shared/links" %>
```
Comment (yuyuvn) [inline app/views/devise/sessions/new.html.erb]: remove

File: app/views/layouts/_header.html.erb [hunk 8]
```diff
@@ -3,7 +3,23 @@
     <%= link_to t(".logo"), root_path, id: "logo" %>
     <nav>
       <ul class="nav navbar-nav navbar-right">
-        <li><%= link_to t(".home"), root_path %></li>
+        <% if user_signed_in? %>
+          <li><%= link_to t(".home"), root_path %></li>
+          <li id="fat-menu" class="dropdown">
+            <a href="#" class="dropdown-toggle" data-toggle="dropdown">
```
Comment (yuyuvn) [inline app/views/layouts/_header.html.erb]: change to link_to

File: app/views/layouts/_header.html.erb [hunk 9]
```diff
@@ -3,7 +3,23 @@
...
+              <li><%= link_to t(".profile"), root_path %></li>
+              <li><%= link_to t(".setting"), root_path %></li>
+              <li class="divider"></li>
+              <li>
+                <%= link_to t(".logout"), sign_out_path%>
+              </li>
+            </ul>
+          </li>
+        <% else %>
+        <% end %>
```
Comment (yuyuvn) [inline app/views/layouts/_header.html.erb]: ??

File: db/migrate/20150706074539_add_devise_to_users.rb [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+      # t.integer  :failed_attempts, default: 0, null: false # Only if lock strategy is :failed_attempts
+      # t.string   :unlock_token # Only if unlock strategy is :email or :both
+      # t.datetime :locked_at
+
+
+      # Uncomment below if timestamps were not included in your original model.
+      # t.timestamps null: false
+    end
+
+    add_index :users, :email,                unique: true
```
Comment (yuyuvn) [inline db/migrate/20150706074539_add_devise_to_users.rb]: convention

File: db/migrate/20150706074539_add_devise_to_users.rb [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    add_index :users, :reset_password_token, unique: true
+    # add_index :users, :confirmation_token,   unique: true
+    # add_index :users, :unlock_token,         unique: true
```
Comment (yuyuvn) [inline db/migrate/20150706074539_add_devise_to_users.rb]: convention

File: app/models/user.rb (hunk 2 above)
Comment (phuongbk) [inline app/models/user.rb]: Enter đó mà chú :3

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (phuongbk) [inline app/controllers/application_controller.rb]: Sure :v:

File: config/routes.rb [hunk 12]
```diff
@@ -1,4 +1,12 @@
 Rails.application.routes.draw do
+
   root "static_pages#home"
 
+  devise_for :users
+
+  devise_scope :user do
```
Comment (dieunb) [inline config/routes.rb]: is it necessary?

**Pull request merged**